security/pkg/nodeagent/caclient/providers/vault/client.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caclient
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/hashicorp/vault/api"
+	"istio.io/istio/pkg/log"
+	caClientInterface "istio.io/istio/security/pkg/nodeagent/caclient/interface"
+)
+
+type vaultClient struct {
+	enableTLS   bool
+	tlsRootCert []byte
+
+	vaultAddr        string
+	vaultLoginRole   string
+	vaultLoginPath   string
+	vaultSignCsrPath string
+
+	client *api.Client
+}
+
+// NewVaultClient create a CA client for the Vault provider 1.
+func NewVaultClient(tls bool, tlsRootCert []byte,
+	vaultAddr, vaultLoginRole, vaultLoginPath, vaultSignCsrPath string) (caClientInterface.Client, error) {
+	c := &vaultClient{
+		enableTLS:        tls,
+		tlsRootCert:      tlsRootCert,
+		vaultAddr:        vaultAddr,
+		vaultLoginRole:   vaultLoginRole,
+		vaultLoginPath:   vaultLoginPath,
+		vaultSignCsrPath: vaultSignCsrPath,
+	}
+
+	var client *api.Client
+	var err error
+	if tls {
+		client, err = createVaultTLSClient(vaultAddr, tlsRootCert)
+	} else {
+		client, err = createVaultClient(vaultAddr)
+	}
+	if err != nil {
+		return nil, err
+	}
+	c.client = client
+
+	return c, nil
+}
+
+// CSR Sign calls Vault to sign a CSR.
+func (c *vaultClient) CSRSign(ctx context.Context, csrPEM []byte, saToken string,
+	certValidTTLInSec int64) ([]string /*PEM-encoded certificate chain*/, error) {
+	token, err := loginVaultK8sAuthMethod(c.client, c.vaultLoginPath, c.vaultLoginRole, saToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to login Vault: %v", err)
+	}
+	c.client.SetToken(token)
+	certChain, err := signCsrByVault(c.client, c.vaultSignCsrPath, certValidTTLInSec, csrPEM)
```

Comment:
csr doesn't contain identity info(which is only available in saToken), both google ca and citadel extract identity from saToken, will vault support the same logic ?